Xamarin.Forms.Core/DataTemplateSelector.cs

```diff
@@ -5,18 +5,20 @@ namespace Xamarin.Forms
 {
 	public abstract class DataTemplateSelector : DataTemplate
 	{
-		Dictionary<Type, DataTemplate> _dataTemplates = new Dictionary<Type, DataTemplate>();
+		readonly Dictionary<Type, DataTemplate> _dataTemplates = new Dictionary<Type, DataTemplate>();
+
+		// NOTE: we don't call base as this is a proxy object that does not need to be bound
+		//       and we do not actually make use of any of the base functionality
+		internal override object OnCreateContent(object item, BindableObject container) =>
+			SelectTemplate(item, container).CreateContent();
 
 		public DataTemplate SelectTemplate(object item, BindableObject container)
 		{
-			var listView = container as ListView;
-
-			var recycle = listView == null ? false :
-				(listView.CachingStrategy & ListViewCachingStrategy.RecycleElementAndDataTemplate) ==
-					ListViewCachingStrategy.RecycleElementAndDataTemplate;
+			var recycle = false;
+			if (container is ListView listView)
+				recycle = listView.CachingStrategy.HasFlag(ListViewCachingStrategy.RecycleElementAndDataTemplate);
 
-			DataTemplate dataTemplate = null;
-			if (recycle && _dataTemplates.TryGetValue(item.GetType(), out dataTemplate))
+			if (recycle && _dataTemplates.TryGetValue(item?.GetType(), out var dataTemplate))
```

Comment:
These null checks on `item` don't really help us, because the result of `item?.GetType()` if `item` is `null` will be `null`, and we'll just get an ArgumentNullException instead of a NullReferenceException. There should probably be a `null` check and exception on `item` at the beginning of the method so it's really obvious why things are failing.

Reply:
```suggestion
			if (recycle && _dataTemplates.TryGetValue(item.GetType(), out var dataTemplate))
```